File: apps/desktop/src/lib/workflow.ts
/**
 * The canvas, as arithmetic (§48, V15).
 *
 * A node editor is mostly geometry, and geometry that is only ever seen is geometry that
 * is never checked. Dragging a box past the edge, dropping a connection on nothing,
 * reordering two actions so the rule runs backwards — none of those look wrong in a
 * screenshot, and all of them change what the rule does.
 *
 * So everything that decides *where a thing is* or *what order it runs in* lives here as a
 * plain function over plain data, and `WorkflowBuilder.tsx` draws the result. The component
 * owns pointer events and nothing else.
 *
 * The shapes mirror `thursday_automation/workflow.py`. Deliberately: the canvas may not be
 * able to draw anything the engine cannot run, so it does not get its own idea of what a
 * workflow is. Validation here is a *convenience* — the server validates the same graph
 * again on `/preview` and on save, and the server's answer is the one that counts.
 */

export type NodeKind = "trigger" | "condition" | "action";

export interface WorkflowNode {
  id: string;
  kind: NodeKind;
  subkind: string;
  config: Record<string, unknown>;
  x: number;
  y: number;
}

export interface Problem {
  node: string;
  message: string;
  severity: "error" | "warning";
}

export interface Consequence {
  node: string;
  kind: string;
  action: string;
  level: string;
  decision: string;
  blocked: boolean;
  unavailable: string;
}

export interface Graph {
  name: string;
  nodes: WorkflowNode[];
  /** Action ids, in the order the engine will run them. */
  order: string[];
  follow_up_from: number | null;
}

/** Box size in canvas units. Connections attach to the edge, so the width is load-bearing. */
export const NODE_WIDTH = 188;
export const NODE_HEIGHT = 62;

/** The canvas is fixed-size rather than infinite: a node dragged out of sight is a node
 *  the owner has lost, and an infinite plane makes that easy and undoable only by luck. */
export const CANVAS = { width: 900, height: 560 };

const COLUMN: Record<NodeKind, number> = { trigger: 40, condition: 300, action: 600 };

/** Clamp a position to the canvas. Applied on every drag, so nothing ever leaves. */
export function clamp(x: number, y: number): { x: number; y: number } {
  return {
    x: Math.min(Math.max(x, 0), CANVAS.width - NODE_WIDTH),
    y: Math.min(Math.max(y, 0), CANVAS.height - NODE_HEIGHT),
  };
}

/** Where a new node of this kind lands: its column, below whatever is already there. */
export function nextPosition(nodes: WorkflowNode[], kind: NodeKind): { x: number; y: number } {
  const column = nodes.filter((n) => n.kind === kind);
  const lowest = column.reduce((best, n) => Math.max(best, n.y + NODE_HEIGHT + 18), 40);
  return clamp(COLUMN[kind], lowest);
}

/** A stable id. Not random: two nodes created in the same millisecond must still differ, and
 *  a test that cannot predict ids cannot assert on them. */
export function nextId(nodes: WorkflowNode[], kind: NodeKind): string {
  const used = new Set(nodes.map((n) => n.id));
  for (let index = 1; ; index += 1) {
    const candidate = `${kind}-${index}`;
    if (!used.has(candidate)) return candidate;
  }
}

export function addNode(graph: Graph, kind: NodeKind, subkind: string): Graph {
  const id = nextId(graph.nodes, kind);
  const { x, y } = nextPosition(graph.nodes, kind);
  const node: WorkflowNode = { id, kind, subkind, config: {}, x, y };
  return {
    ...graph,
    nodes: [...graph.nodes, node],
    // An action is only in the rule if it is in the order, so adding one appends it.
    order: kind === "action" ? [...graph.order, id] : graph.order,
  };
}

export function removeNode(graph: Graph, id: string): Graph {
  const order = graph.order.filter((n) => n !== id);
  const remaining = graph.nodes.filter((n) => n.id !== id);
  const actions = remaining.filter((n) => n.kind === "action").length;
  return {
    ...graph,
    nodes: remaining,
    order,
    // A split that now points past the end would silently turn a follow-up into a main
    // action, or the reverse. Cleared rather than guessed.
    follow_up_from:
      graph.follow_up_from !== null && graph.follow_up_from <= actions && graph.follow_up_from > 0
        ? graph.follow_up_from
        : null,
  };
}

export function moveNode(graph: Graph, id: string, x: number, y: number): Graph {
  const at = clamp(x, y);
  return { ...graph, nodes: graph.nodes.map((n) => (n.id === id ? { ...n, ...at } : n)) };
}

export function setConfig(graph: Graph, id: string, config: Record<string, unknown>): Graph {
  return { ...graph, nodes: graph.nodes.map((n) => (n.id === id ? { ...n, config } : n)) };
}

export function setSubkind(graph: Graph, id: string, subkind: string): Graph {
  // The config belongs to the old subkind: a `cron` left on an event trigger would be sent
  // to the server and silently ignored, and the owner would never know which one was live.
  return {
    ...graph,
    nodes: graph.nodes.map((n) => (n.id === id ? { ...n, subkind, config: {} } : n)),
  };
}

/**
 * Move an action earlier or later in the run order.
 *
 * The order *is* the rule — actions run in it — so reordering is an edit to behaviour, not
 * to the picture. A drag that reorders by position would make the rule depend on where the
 * owner happened to drop a box; the list is explicit instead.
 */
export function reorder(graph: Graph, id: string, direction: -1 | 1): Graph {
  const from = graph.order.indexOf(id);
  const to = from + direction;
  if (from < 0 || to < 0 || to >= graph.order.length) return graph;
  const order = [...graph.order];
  [order[from], order[to]] = [order[to], order[from]];
  return { ...graph, order };
}

/** Actions in run order. Anything the order forgot still appears, at the end — mirroring
 *  the server, where dropping it would make a box vanish from the rule but not the canvas. */
export function orderedActions(graph: Graph): WorkflowNode[] {
  const actions = graph.nodes.filter((n) => n.kind === "action");
  const byId = new Map(actions.map((n) => [n.id, n]));
  const listed = graph.order.map((id) => byId.get(id)).filter((n): n is WorkflowNode => !!n);
  const seen = new Set(listed.map((n) => n.id));
  return [...listed, ...actions.filter((n) => !seen.has(n.id))];
}

export interface Edge {
  from: string;
  to: string;
}

/**
 * The wires. Derived, never stored.
 *
 * The engine's shape is fixed — one trigger, an AND-set of conditions, then actions in
 * order — so there is nothing for the owner to connect and nothing they could connect
 * wrongly. Letting them draw arrows would mean letting them draw one the engine ignores,
 * which is the failure this whole module exists to avoid (ADR 0064).
 */
export function edges(graph: Graph): Edge[] {
  const trigger = graph.nodes.find((n) => n.kind === "trigger");
  if (!trigger) return [];
  const conditions = graph.nodes.filter((n) => n.kind === "condition");
  const actions = orderedActions(graph);

  const out: Edge[] = conditions.map((c) => ({ from: trigger.id, to: c.id }));
  const heads = conditions.length ? conditions.map((c) => c.id) : [trigger.id];
  if (actions.length) {
    for (const head of heads) out.push({ from: head, to: actions[0].id });
    for (let i = 1; i < actions.length; i += 1) {
      out.push({ from: actions[i - 1].id, to: actions[i].id });
    }
  }
  return out;
}

/** A cubic curve between two boxes, right edge to left edge. */
export function curve(from: WorkflowNode, to: WorkflowNode): string {
  const x1 = from.x + NODE_WIDTH;
  const y1 = from.y + NODE_HEIGHT / 2;
  const x2 = to.x;
  const y2 = to.y + NODE_HEIGHT / 2;
  const reach = Math.max(36, Math.abs(x2 - x1) / 2);
  return `M ${x1} ${y1} C ${x1 + reach} ${y1}, ${x2 - reach} ${y2}, ${x2} ${y2}`;
}

/**
 * Local validation, so the canvas can mark a box before the round trip.
 *
 * Never the last word: `/preview` runs the server's own validator, and anything only it
 * knows — a cron it cannot parse, a tool that does not exist, a handler that is not wired —
 * arrives from there. This function exists to make typing feel responsive, not to decide
 * whether a rule is safe.
 */
export function localProblems(graph: Graph): Problem[] {
  const problems: Problem[] = [];
  const triggers = graph.nodes.filter((n) => n.kind === "trigger");
  if (!graph.name.trim()) problems.push({ node: "", message: "กฎต้องมีชื่อ", severity: "error" });
  if (triggers.length === 0) {
    problems.push({ node: "", message: "ต้องมีตัวกระตุ้น (WHEN) หนึ่งอัน", severity: "error" });
  } else if (triggers.length > 1) {
    problems.push({
      node: "",
      message: `มีตัวกระตุ้น ${triggers.length} อัน — เครื่องมือนี้รับได้อันเดียว`,
      severity: "error",
    });
  }
  if (!graph.nodes.some((n) => n.kind === "action")) {
    problems.push({
      node: "",
      message: "ต้องมีการกระทำ (DO) อย่างน้อยหนึ่งอย่าง",
      severity: "error",
    });
  }
  return problems;
}

export function hasErrors(problems: Problem[]): boolean {
  return problems.some((p) => p.severity === "error");
}

export function problemsFor(problems: Problem[], id: string): Problem[] {
  return problems.filter((p) => p.node === id);
}

export const EMPTY: Graph = { name: "", nodes: [], order: [], follow_up_from: null };

File: packages/automation/thursday_automation/workflow.py
"""The node graph behind the workflow builder (§48, V15).

A visual builder has one failure mode that matters, and it is not a missing feature. It is
a canvas that lets the owner draw something the engine cannot run, or can run but not the
way the picture reads. Then the rule is saved, looks right, and does something else — and
the owner finds out at the moment the rule mattered.

So the graph here is **the engine's shape, drawn**, not a general dataflow language:

    [trigger] ──▶ [condition] ──▶ [condition] ──▶ [action] ──▶ [action]
                   all must hold                  in order, then follow-ups

Exactly one trigger. Conditions are an AND-set — the engine evaluates `all(...)`, so there
is no OR node, no branch and no loop, because drawing one would be drawing a promise.
Actions run in order, and `follow_ups` is the tail of the same list marked as such.

Three things this module does that the canvas alone could not:

**It converts, both ways.** `to_automation` and `from_automation` round-trip, so a rule
written by `routines.py` or restored from disk opens in the builder instead of being
invisible to it.

**It refuses rather than corrects.** `validate` returns every problem with the node that
caused it. A graph with a problem does not save; nothing here quietly drops a dangling node
or invents a missing field.

**It says what would happen, before it happens.** `preview` reports, per action, what the
Permission Engine would decide and whether a handler for that kind of action is even wired
up — the unwired case being the one a picture hides best (ADR 0064).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal
from uuid import UUID

from thursday_shared.enums import NotificationPriority, PermissionLevel, PolicyDecision
from thursday_shared.ids import new_id
from thursday_shared.models import Budget

from thursday_automation.cron import CronRefused, describe, parse
from thursday_automation.rules import Action, Automation, Condition, Trigger

NodeKind = Literal["trigger", "condition", "action"]

#: Trigger kinds the builder offers, and what actually runs them. A kind with no runner is
#: listed here with the reason rather than left off the menu silently — the owner asking
#: "why can't I do X" deserves the answer, and a future runner turns one line green.
TRIGGER_RUNNERS: dict[str, str] = {
    "event": "",
    "schedule": "",
    "manual": "",
    "state_change": (
        "ยังไม่มีตัวเฝ้าดูสถานะ — ไม่มีโค้ดส่วนไหนคอยดูว่าค่าใน world state เปลี่ยน กฎที่ใช้ตัวกระตุ้นนี้จะไม่ทำงานเลย"
    ),
}

#: Action kinds, and the container attribute each needs in order to do anything. The engine
#: returns `{"skipped": kind}` when the attribute is absent, which from the outside looks
#: exactly like a rule that ran and found nothing to do.
ACTION_NEEDS: dict[str, str] = {
    "notify": "",
    "task": "tasks",
    "tool": "executor",
    "obsidian_write": "executor",
}

CONDITION_OPS: tuple[str, ...] = ("eq", "ne", "gt", "lt", "contains", "matches", "in")


class WorkflowRefused(ValueError):
    """A graph this module will not turn into a rule."""


@dataclass
class Node:
    """One box on the canvas. `x`/`y` are the owner's layout and mean nothing to the engine."""

    id: str
    kind: NodeKind
    #: For a trigger: the trigger kind. For a condition: the operator. For an action: the
    #: action kind. Always the thing that decides how `config` is read.
    subkind: str = ""
    config: dict[str, Any] = field(default_factory=dict)
    x: float = 0.0
    y: float = 0.0

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "kind": self.kind,
            "subkind": self.subkind,
            "config": dict(self.config),
            "x": self.x,
            "y": self.y,
        }


@dataclass
class Problem:
    """Something wrong, and which box it is in. `node` is empty for a whole-graph problem."""

    node: str
    message: str
    #: "error" stops the rule saving. "warning" does not, and must never be used for
    #: something that makes the rule not run — that is an error wearing a softer word.
    severity: Literal["error", "warning"] = "error"

    def to_dict(self) -> dict[str, Any]:
        return {"node": self.node, "message": self.message, "severity": self.severity}


@dataclass
class Workflow:
    """A graph, and the rule it means."""

    name: str = ""
    nodes: list[Node] = field(default_factory=list)
    #: Ordered: the action order on the canvas is the order the engine runs them in.
    order: list[str] = field(default_factory=list)
    #: Where the follow-up tail starts within the action nodes. `None` means no follow-ups.
    follow_up_from: int | None = None
    enabled: bool = False
    automation_id: UUID | None = None
    priority: NotificationPriority = NotificationPriority.NORMAL
    budget: Budget = field(default_factory=lambda: Budget(usd=0.20, seconds=120, tool_calls=10))

    # ----------------------------------------------------------------- shape

    @property
    def triggers(self) -> list[Node]:
        return [n for n in self.nodes if n.kind == "trigger"]

    @property
    def conditions(self) -> list[Node]:
        return [n for n in self.nodes if n.kind == "condition"]

    @property
    def actions(self) -> list[Node]:
        """Action nodes in the order the owner arranged them."""
        by_id = {n.id: n for n in self.nodes if n.kind == "action"}
        ordered = [by_id.pop(node_id) for node_id in self.order if node_id in by_id]
        # Anything the order forgot still appears, at the end. Dropping it would make a
        # node vanish from the rule while staying on the canvas.
        return [*ordered, *by_id.values()]

    # ------------------------------------------------------------ validation

    def validate(self) -> list[Problem]:
        """Every reason this graph is not a rule. Empty means it is."""
        problems: list[Problem] = []

        if not self.name.strip():
            problems.append(Problem("", "กฎต้องมีชื่อ"))

        triggers = self.triggers
        if not triggers:
            problems.append(Problem("", "ต้องมีตัวกระตุ้น (WHEN) หนึ่งอัน"))
        elif len(triggers) > 1:
            # The engine has one `trigger` field. Two on the canvas would mean one is
            # silently ignored, and the picture would not say which.
            problems.append(Problem("", f"มีตัวกระตุ้น {len(triggers)} อัน — เครื่องมือนี้รับได้อันเดียว"))

        for node in triggers:
            problems.extend(self._trigger_problems(node))
        for node in self.conditions:
            problems.extend(self._condition_problems(node))

        actions = self.actions
        if not actions:
            problems.append(Problem("", "ต้องมีการกระทำ (DO) อย่างน้อยหนึ่งอย่าง"))
        for node in actions:
            problems.extend(self._action_problems(node))

        if self.follow_up_from is not None and not 0 < self.follow_up_from <= len(actions):
            problems.append(
                Problem("", f"จุดเริ่ม follow-up ({self.follow_up_from}) อยู่นอกลำดับการกระทำ")
            )

        return problems

    def _trigger_problems(self, node: Node) -> list[Problem]:
        if node.subkind not in TRIGGER_RUNNERS:
            return [Problem(node.id, f"ไม่รู้จักตัวกระตุ้นชนิด {node.subkind!r}")]

        unavailable = TRIGGER_RUNNERS[node.subkind]
        if unavailable:
            return [Problem(node.id, unavailable)]

        if node.subkind == "schedule":
            try:
                parse(str(node.config.get("cron") or ""))
            except CronRefused as exc:
                return [Problem(node.id, str(exc))]
        if node.subkind == "event" and not str(node.config.get("event_kind") or "").strip():
            return [Problem(node.id, "ตัวกระตุ้นแบบเหตุการณ์ต้องระบุชนิดเหตุการณ์")]
        return []

    def _condition_problems(self, node: Node) -> list[Problem]:
        problems: list[Problem] = []
        if node.subkind not in CONDITION_OPS:
            problems.append(Problem(node.id, f"ไม่รู้จักตัวเปรียบเทียบ {node.subkind!r}"))
        if not str(node.config.get("field") or "").strip():
            problems.append(Problem(node.id, "เงื่อนไขต้องระบุฟิลด์ที่จะตรวจ"))
        if node.subkind in ("gt", "lt"):
            try:
                float(node.config.get("value"))  # type: ignore[arg-type]
            except (TypeError, ValueError):
                # `gt` on a non-number compares NaN, which is False for everything — a
                # condition that never holds, and a rule that never fires, with no error.
                problems.append(
                    Problem(
                        node.id, f"{node.subkind} ต้องเทียบกับตัวเลข — ได้ {node.config.get('value')!r}"
                    )
                )
        if node.subkind == "in" and not isinstance(node.config.get("value"), list):
            problems.append(Problem(node.id, "in ต้องเทียบกับรายการ"))
        return problems

    def _action_problems(self, node: Node) -> list[Problem]:
        if node.subkind not in ACTION_NEEDS:
            return [Problem(node.id, f"ไม่รู้จักการกระทำชนิด {node.subkind!r}")]
        if node.subkind == "tool" and not str(node.config.get("name") or "").strip():
            return [Problem(node.id, "การกระทำแบบเครื่องมือต้องระบุชื่อเครื่องมือ")]
        if node.subkind == "notify" and not str(node.config.get("title") or "").strip():
            return [Problem(node.id, "การแจ้งเตือนต้องมีหัวข้อ", "warning")]
        return []

    # ------------------------------------------------------------ conversion

    def to_automation(self) -> Automation:
        """The rule this graph means. Refuses a graph that has errors."""
        errors = [p for p in self.validate() if p.severity == "error"]
        if errors:
            raise WorkflowRefused(" / ".join(p.message for p in errors))

        trigger_node = self.triggers[0]
        trigger = Trigger(kind=trigger_node.subkind)  # type: ignore[arg-type]
        if trigger.kind == "event":
            trigger.event_kind = str(trigger_node.config.get("event_kind") or "*")
        elif trigger.kind == "schedule":
            trigger.cron = str(trigger_node.config.get("cron"))
        elif trigger.kind == "state_change":  # pragma: no cover - validate refuses first
            trigger.field = str(trigger_node.config.get("field") or "")

        actions = self.actions
        split = self.follow_up_from if self.follow_up_from is not None else len(actions)

        return Automation(
            id=self.automation_id or new_id(),
            name=self.name.strip(),
            trigger=trigger,
            conditions=[
                Condition(
                    field=str(n.config.get("field")),
                    op=n.subkind,  # type: ignore[arg-type]
                    value=n.config.get("value"),
                )
                for n in self.conditions
            ],
            actions=[_action(n) for n in actions[:split]],
            follow_ups=[_action(n) for n in actions[split:]],
            enabled=self.enabled,
            priority=self.priority,
            budget=self.budget,
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "nodes": [n.to_dict() for n in self.nodes],
            "order": list(self.order),
            "follow_up_from": self.follow_up_from,
            "enabled": self.enabled,
            "automation_id": str(self.automation_id) if self.automation_id else None,
            "priority": str(self.priority),
            "count": len(self.nodes),
            "items": [n.to_dict() for n in self.nodes],
        }


def _action(node: Node) -> Action:
    """An action node's config is `name` plus everything else, which becomes `args`.

    Flat rather than nested, because the canvas edits fields and a nested `args` object
    would need an editor of its own — and a field the owner cannot see is a field that
    quietly keeps whatever the last person put in it.
    """
    config = dict(node.config)
    return Action(
        kind=node.subkind,  # type: ignore[arg-type]
        name=str(config.pop("name", "")),
        args=config,
        device_hint=None,
    )


def from_automation(
    automation: Automation, *, layout: dict[str, tuple[float, float]] | None = None
) -> Workflow:
    """Open an existing rule on the canvas.

    Rules written by `routines.py`, restored from disk, or suggested by Thursday are all
    ordinary automations. Without this they would be invisible to the builder, and the
    owner would have two places rules live — one they can see and one they cannot.
    """
    places = layout or {}
    nodes: list[Node] = []
    order: list[str] = []

    def place(node_id: str, column: int, row: int) -> tuple[float, float]:
        return places.get(node_id, (120.0 + column * 240.0, 80.0 + row * 110.0))

    trigger_id = "trigger"
    config: dict[str, Any] = {}
    if automation.trigger.kind == "event":
        config["event_kind"] = automation.trigger.event_kind
    elif automation.trigger.kind == "schedule":
        config["cron"] = automation.trigger.cron or ""
    elif automation.trigger.kind == "state_change":
        config["field"] = automation.trigger.field or ""
    x, y = place(trigger_id, 0, 0)
    nodes.append(Node(trigger_id, "trigger", automation.trigger.kind, config, x, y))

    for index, condition in enumerate(automation.conditions):
        node_id = f"condition-{index}"
        x, y = place(node_id, 1, index)
        nodes.append(
            Node(
                node_id,
                "condition",
                condition.op,
                {"field": condition.field, "value": condition.value},
                x,
                y,
            )
        )

    every = [*automation.actions, *automation.follow_ups]
    for index, action in enumerate(every):
        node_id = f"action-{index}"
        x, y = place(node_id, 2, index)
        nodes.append(
            Node(node_id, "action", action.kind, {"name": action.name, **action.args}, x, y)
        )
        order.append(node_id)

    return Workflow(
        name=automation.name,
        nodes=nodes,
        order=order,
        follow_up_from=len(automation.actions) if automation.follow_ups else None,
        enabled=automation.enabled,
        automation_id=automation.id,
        priority=automation.priority,
        budget=automation.budget,
    )


# ------------------------------------------------------------------------ preview


@dataclass
class Consequence:
    """What one action node would actually do, if the rule fired right now."""

    node: str
    kind: str
    #: The permission verb this resolves to, as the Permission Engine would see it.
    action: str
    level: PermissionLevel
    decision: PolicyDecision
    blocked: bool
    #: Empty when the action can run. Otherwise why it cannot, in the owner's words.
    unavailable: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "node": self.node,
            "kind": self.kind,
            "action": self.action,
            # `.name`, not `str()`: PermissionLevel is an IntEnum, so str() gives "3" and
            # the screen would show a number nobody can read.
            "level": self.level.name,
            "decision": str(self.decision),
            "blocked": self.blocked,
            "unavailable": self.unavailable,
        }


def preview(
    workflow: Workflow, *, policies: Any, wired: set[str] | None = None
) -> list[Consequence]:
    """What each action would do — permission decision first, then whether it can run at all.

    `wired` names the container attributes that are actually present. An action whose
    handler is missing returns `{"skipped": ...}` from the engine, which is indistinguishable
    from a rule that ran and had nothing to do. Saying so here, while the owner is looking
    at the canvas, is the only moment it is cheap to say.
    """
    present = wired if wired is not None else set(ACTION_NEEDS.values())
    out: list[Consequence] = []

    for node in workflow.actions:
        verb = _verb(node)
        policy = policies.get(verb)
        needs = ACTION_NEEDS.get(node.subkind, "")
        unavailable = ""
        if needs and needs not in present:
            unavailable = f"ยังไม่ได้ต่อ {needs} — การกระทำนี้จะถูกข้ามไปเงียบ ๆ"
        out.append(
            Consequence(
                node=node.id,
                kind=node.subkind,
                action=verb,
                level=policy.level,
                decision=policy.default,
                blocked=bool(getattr(policies, "is_blocked", lambda _n: False)(verb)),
                unavailable=unavailable,
            )
        )
    return out


def _verb(node: Node) -> str:
    """The permission verb an action node resolves to."""
    if node.subkind == "tool":
        return str(node.config.get("name") or "")
    if node.subkind == "obsidian_write":
        return "obsidian.write"
    if node.subkind == "notify":
        return "notify.show"
    return "task.create"


def explain(workflow: Workflow) -> str:
    """One line the owner can check against what they meant. No graph, no jargon."""
    trigger = workflow.triggers[0] if workflow.triggers else None
    if trigger is None:
        return "ยังไม่มีตัวกระตุ้น"

    if trigger.subkind == "schedule":
        try:
            when = describe(str(trigger.config.get("cron") or ""))
        except CronRefused:
            when = "ตารางเวลาที่อ่านไม่ออก"
    elif trigger.subkind == "event":
        when = f"เมื่อเกิด {trigger.config.get('event_kind') or '*'}"
    elif trigger.subkind == "manual":
        when = "เมื่อสั่งเอง"
    else:
        when = f"เมื่อ {trigger.subkind}"

    parts = [when]
    if workflow.conditions:
        parts.append(f"ถ้าเข้าเงื่อนไขครบทั้ง {len(workflow.conditions)} ข้อ")
    actions = workflow.actions
    if actions:
        parts.append("แล้ว " + " → ".join(_name(n) for n in actions))
    return " ".join(parts)


def _name(node: Node) -> str:
    if node.subkind == "tool":
        return str(node.config.get("name") or "เครื่องมือ")
    return {"notify": "แจ้งเตือน", "task": "สร้างงาน", "obsidian_write": "บันทึกลง Obsidian"}.get(
        node.subkind, node.subkind
    )
